File: typescript/07-estado-contexto/persistencia.ts
// Cómo ejecutar: make ts SCRIPT=typescript/07-estado-contexto/persistencia.ts
import { Database } from "bun:sqlite";
import { randomUUID } from "crypto";

const SCHEMA_VERSION = 1;

enum CheckpointTrigger {
  PHASE_COMPLETE   = "phase_complete",
  IRREVERSIBLE_ACT = "irreversible_act",
  USER_REQUEST     = "user_request",
  BUDGET_THRESHOLD = "budget_threshold",
  PERIODIC         = "periodic",
}

interface CheckpointData {
  id: string;
  task_id: string;
  messages: object[];
  trigger: string;
  schema_version: number;
  metadata: Record<string, unknown>;
  timestamp: number;
}

class Checkpoint implements CheckpointData {
  id: string;
  task_id: string;
  messages: object[];
  trigger: string;
  schema_version: number;
  metadata: Record<string, unknown>;
  timestamp: number;

  constructor(data: Partial<CheckpointData> & { task_id: string; messages: object[] }) {
    this.id             = data.id             ?? randomUUID();
    this.task_id        = data.task_id;
    this.messages       = data.messages;
    this.trigger        = data.trigger        ?? CheckpointTrigger.PHASE_COMPLETE;
    this.schema_version = data.schema_version ?? SCHEMA_VERSION;
    this.metadata       = data.metadata       ?? {};
    this.timestamp      = data.timestamp      ?? Date.now() / 1000;
  }

  toJSON(): string {
    return JSON.stringify({
      id: this.id, task_id: this.task_id, messages: this.messages,
      trigger: this.trigger, schema_version: this.schema_version,
      metadata: this.metadata, timestamp: this.timestamp,
    });
  }

  static fromJSON(data: string): Checkpoint {
    return new Checkpoint(JSON.parse(data));
  }
}

abstract class CheckpointStore {
  abstract save(checkpoint: Checkpoint): string;
  abstract load(checkpointId: string): Checkpoint | null;
  abstract list(taskId: string): Checkpoint[];
  abstract latest(taskId: string): Checkpoint | null;
}

class SQLiteCheckpointStore extends CheckpointStore {
  private db: Database;

  constructor(dbPath = ":memory:") {
    super();
    this.db = new Database(dbPath);
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS checkpoints (
        id             TEXT PRIMARY KEY,
        task_id        TEXT NOT NULL,
        schema_version INTEGER NOT NULL,
        trigger        TEXT NOT NULL,
        timestamp      REAL NOT NULL,
        data           TEXT NOT NULL
      )
    `);
    this.db.exec(
      "CREATE INDEX IF NOT EXISTS idx_task ON checkpoints(task_id, timestamp)"
    );
  }

  save(checkpoint: Checkpoint): string {
    this.db
      .prepare("INSERT INTO checkpoints VALUES (?,?,?,?,?,?)")
      .run(
        checkpoint.id,
        checkpoint.task_id,
        checkpoint.schema_version,
        checkpoint.trigger,
        checkpoint.timestamp,
        checkpoint.toJSON()
      );
    return checkpoint.id;
  }

  load(checkpointId: string): Checkpoint | null {
    const row = this.db
      .prepare("SELECT data, schema_version FROM checkpoints WHERE id=?")
      .get(checkpointId) as { data: string; schema_version: number } | undefined;
    if (!row) return null;
    let cp = Checkpoint.fromJSON(row.data);
    if (cp.schema_version !== SCHEMA_VERSION) {
      cp = downgradeCheckpoint(cp, SCHEMA_VERSION);
    }
    return cp;
  }

  list(taskId: string): Checkpoint[] {
    const rows = this.db
      .prepare("SELECT data FROM checkpoints WHERE task_id=? ORDER BY timestamp ASC")
      .all(taskId) as { data: string }[];
    return rows.map((r) => Checkpoint.fromJSON(r.data));
  }

  latest(taskId: string): Checkpoint | null {
    const row = this.db
      .prepare("SELECT data FROM checkpoints WHERE task_id=? ORDER BY timestamp DESC LIMIT 1")
      .get(taskId) as { data: string } | undefined;
    return row ? Checkpoint.fromJSON(row.data) : null;
  }
}

function downgradeCheckpoint(checkpoint: Checkpoint, targetVersion: number): Checkpoint {
  if (checkpoint.schema_version === targetVersion) return checkpoint;
  return new Checkpoint({
    task_id:        checkpoint.task_id,
    messages:       checkpoint.messages,
    trigger:        "downgraded_from_v" + checkpoint.schema_version,
    schema_version: targetVersion,
    metadata:       { downgraded: true, original_version: checkpoint.schema_version },
    id:             checkpoint.id,
    timestamp:      checkpoint.timestamp,
  });
}

function shouldCheckpoint(trigger: string, toolCallsSinceLast = 0): boolean {
  const primaryTriggers = new Set([
    CheckpointTrigger.PHASE_COMPLETE,
    CheckpointTrigger.IRREVERSIBLE_ACT,
    CheckpointTrigger.USER_REQUEST,
  ]);
  if (primaryTriggers.has(trigger as CheckpointTrigger)) return true;
  if (trigger === CheckpointTrigger.PERIODIC) return toolCallsSinceLast >= 20;
  return false;
}

const store = new SQLiteCheckpointStore();
const task = "analisis-repo-facturacion";

const msgs = [
  { role: "user",      content: "Analiza el módulo de auth." },
  { role: "assistant", content: "Encontré 2 vulnerabilidades en auth.py." },
];

const cp1 = new Checkpoint({
  task_id:  task,
  messages: msgs,
  trigger:  CheckpointTrigger.PHASE_COMPLETE,
  metadata: { fase: 1, vulnerabilidades: 2 },
});
const id1 = store.save(cp1);
console.log(`Guardado: ${id1.slice(0, 8)}... | trigger=${cp1.trigger}`);

const cargado = store.load(id1)!;
console.log(`Cargado: task=${cargado.task_id} | msgs=${cargado.messages.length} | v=${cargado.schema_version}`);

const cpViejo = new Checkpoint({ task_id: task, messages: msgs, schema_version: 0 });
const cpDegradado = downgradeCheckpoint(cpViejo, SCHEMA_VERSION);
console.log(`Degradado: v${cpViejo.schema_version}→v${cpDegradado.schema_version} | ${JSON.stringify(cpDegradado.metadata)}`);

console.log(`\nshould_checkpoint(phase_complete): ${shouldCheckpoint("phase_complete")}`);
console.log(`should_checkpoint(periodic, 5 tool calls): ${shouldCheckpoint("periodic", 5)}`);
console.log(`should_checkpoint(periodic, 25 tool calls): ${shouldCheckpoint("periodic", 25)}`);

const todos = store.list(task);
console.log(`\nCheckpoints para '${task}': ${todos.length}`);

File: python/07-estado-contexto/persistencia.py
"""Checkpointing del estado del agente con SQLite.

Checkpoint: snapshot del estado en un punto semántico.
SQLiteCheckpointStore: implementación concreta con interfaz abstracta.
schema_version: previene deserialización incorrecta cuando el agente evoluciona.
downgrade_checkpoint: convierte un checkpoint de versión antigua al formato actual.
should_checkpoint: decide qué eventos justifican un checkpoint (semánticos, no mecánicos).

Cómo ejecutar:
    make py SCRIPT=python/07-estado-contexto/persistencia.py

Qué esperar:
    Demo de checkpointing en SQLite: crear, guardar, restaurar y migrar
    checkpoints entre versiones del agente. Los checkpoints son semanticos,
    no mecanicos (no se guarda en cada iteracion sino en puntos clave).

Variables de entorno:
    MODEL — modelo a usar (default: claude-sonnet-4-6)
"""
import json
import sqlite3
import time
import uuid
from abc import ABC, abstractmethod
from dataclasses import asdict, dataclass, field
from enum import Enum
from typing import Optional


SCHEMA_VERSION = 1  # incrementar en cada cambio estructural del Checkpoint


class CheckpointTrigger(Enum):
    PHASE_COMPLETE   = "phase_complete"    # el agente completó una fase del plan
    IRREVERSIBLE_ACT = "irreversible_act"  # acción que no se puede deshacer
    USER_REQUEST     = "user_request"      # el usuario pidió pausar
    BUDGET_THRESHOLD = "budget_threshold"  # contexto llegó al umbral de compactación
    PERIODIC         = "periodic"          # fallback si han pasado N tool calls sin checkpoint


@dataclass
class Checkpoint:
    task_id:        str
    messages:       list[dict]
    trigger:        str  = CheckpointTrigger.PHASE_COMPLETE.value
    schema_version: int  = SCHEMA_VERSION
    metadata:       dict = field(default_factory=dict)  # estado específico de la app
    id:             str  = field(default_factory=lambda: str(uuid.uuid4()))
    timestamp:      float = field(default_factory=time.time)

    def to_json(self) -> str:
        return json.dumps(asdict(self), ensure_ascii=False)

    @classmethod
    def from_json(cls, data: str) -> "Checkpoint":
        return cls(**json.loads(data))


class CheckpointStore(ABC):
    @abstractmethod
    def save(self, checkpoint: Checkpoint) -> str: ...

    @abstractmethod
    def load(self, checkpoint_id: str) -> Optional[Checkpoint]: ...

    @abstractmethod
    def list(self, task_id: str) -> list[Checkpoint]: ...

    @abstractmethod
    def latest(self, task_id: str) -> Optional[Checkpoint]: ...


class SQLiteCheckpointStore(CheckpointStore):
    def __init__(self, db_path: str = ":memory:"):
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.execute("""
            CREATE TABLE IF NOT EXISTS checkpoints (
                id             TEXT PRIMARY KEY,
                task_id        TEXT NOT NULL,
                schema_version INTEGER NOT NULL,
                trigger        TEXT NOT NULL,
                timestamp      REAL NOT NULL,
                data           TEXT NOT NULL
            )
        """)
        self.conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_task ON checkpoints(task_id, timestamp)"
        )
        self.conn.commit()

    def save(self, checkpoint: Checkpoint) -> str:
        self.conn.execute(
            "INSERT INTO checkpoints VALUES (?,?,?,?,?,?)",
            (
                checkpoint.id,
                checkpoint.task_id,
                checkpoint.schema_version,
                checkpoint.trigger,
                checkpoint.timestamp,
                checkpoint.to_json(),
            ),
        )
        self.conn.commit()
        return checkpoint.id

    def load(self, checkpoint_id: str) -> Optional[Checkpoint]:
        row = self.conn.execute(
            "SELECT data, schema_version FROM checkpoints WHERE id=?",
            (checkpoint_id,),
        ).fetchone()
        if not row:
            return None
        data, stored_version = row
        cp = Checkpoint.from_json(data)
        if cp.schema_version != SCHEMA_VERSION:
            cp = downgrade_checkpoint(cp, SCHEMA_VERSION)
        return cp

    def list(self, task_id: str) -> list[Checkpoint]:
        rows = self.conn.execute(
            "SELECT data FROM checkpoints WHERE task_id=? ORDER BY timestamp ASC",
            (task_id,),
        ).fetchall()
        return [Checkpoint.from_json(r[0]) for r in rows]

    def latest(self, task_id: str) -> Optional[Checkpoint]:
        row = self.conn.execute(
            "SELECT data FROM checkpoints WHERE task_id=? ORDER BY timestamp DESC LIMIT 1",
            (task_id,),
        ).fetchone()
        return Checkpoint.from_json(row[0]) if row else None


def downgrade_checkpoint(checkpoint: Checkpoint, target_version: int) -> Checkpoint:
    """Convierte un checkpoint a target_version ante incompatibilidad de schema.

    En lugar de fallar, produce un checkpoint degradado con el que el agente
    puede continuar con contexto parcial. En producción: añadir cases por versión.
    """
    if checkpoint.schema_version == target_version:
        return checkpoint

    # Degradación universal: preservar messages y task_id, descartar metadata específica
    return Checkpoint(
        task_id=checkpoint.task_id,
        messages=checkpoint.messages,
        trigger="downgraded_from_v" + str(checkpoint.schema_version),
        schema_version=target_version,
        metadata={
            "downgraded": True,
            "original_version": checkpoint.schema_version,
        },
        id=checkpoint.id,
        timestamp=checkpoint.timestamp,
    )


def should_checkpoint(trigger: str, tool_calls_since_last: int = 0) -> bool:
    """Decide si el evento actual justifica un checkpoint.

    PHASE_COMPLETE e IRREVERSIBLE_ACT siempre hacen checkpoint.
    PERIODIC solo si llevamos ≥20 tool calls sin checkpoint (fallback de seguridad).
    """
    primary_triggers = {
        CheckpointTrigger.PHASE_COMPLETE.value,
        CheckpointTrigger.IRREVERSIBLE_ACT.value,
        CheckpointTrigger.USER_REQUEST.value,
    }
    if trigger in primary_triggers:
        return True
    if trigger == CheckpointTrigger.PERIODIC.value:
        return tool_calls_since_last >= 20
    return False


if __name__ == "__main__":
    store = SQLiteCheckpointStore()
    task = "analisis-repo-facturacion"

    # Guardar checkpoint al completar la primera fase
    msgs = [
        {"role": "user", "content": "Analiza el módulo de auth."},
        {"role": "assistant", "content": "Encontré 2 vulnerabilidades en auth.py."},
    ]
    cp1 = Checkpoint(
        task_id=task,
        messages=msgs,
        trigger=CheckpointTrigger.PHASE_COMPLETE.value,
        metadata={"fase": 1, "vulnerabilidades": 2},
    )
    id1 = store.save(cp1)
    print(f"Guardado: {id1[:8]}... | trigger={cp1.trigger}")

    # Cargar y verificar
    cargado = store.load(id1)
    print(f"Cargado: task={cargado.task_id} | msgs={len(cargado.messages)} | v={cargado.schema_version}")

    # Simular checkpoint de versión antigua
    cp_viejo = Checkpoint(task_id=task, messages=msgs, schema_version=0)
    cp_degradado = downgrade_checkpoint(cp_viejo, SCHEMA_VERSION)
    print(f"Degradado: v{cp_viejo.schema_version}→v{cp_degradado.schema_version} | {cp_degradado.metadata}")

    # Verificar decisión de checkpoint
    print(f"\nshould_checkpoint(phase_complete): {should_checkpoint('phase_complete')}")
    print(f"should_checkpoint(periodic, 5 tool calls): {should_checkpoint('periodic', 5)}")
    print(f"should_checkpoint(periodic, 25 tool calls): {should_checkpoint('periodic', 25)}")

    # Listar todos los checkpoints de la tarea
    todos = store.list(task)
    print(f"\nCheckpoints para '{task}': {len(todos)}")
